Sprint 15

Kolesnikov-Eugene opened the pull request.

Kolesnikov-Eugene committed:
created dataManager, DataStrore
TrackerCoreData, TrackerCategoryCoreData, TrackerRecordCoreData
all trackers and categories stored in CoreData + completedTrackers strored
fetchRequests for categories and trackers completed

Kolesnikov-Eugene committed:
NSFetchedResultsController updates collectionView after adding a new tracker to DB.
Changed logic of adding new trackers to DB

Kolesnikov-Eugene committed:
Removed force unwrapping in DataStore

So1ovyov commented on Tracker/AppDelegate.swift:
```diff
@@ -30,6 +31,23 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        return container
+    }()
+    
+//    lazy var trackerCategoryDataStore = {
```
Закомментированный код можно удалить, если он не используется.

Kolesnikov-Eugene replied:
да, упустил из виду(
Почистил весь неиспользуемый код в проекте.

So1ovyov commented on Tracker/Domain/DataManager.swift:
```diff
@@ -0,0 +1,136 @@
+    }
+    
+    func deleteTrackerRecord(_ trackerID: UUID, for date: Date) throws {
+        //TODO
```
Давай подпишем что именно в дальнейшем ты тут хочешь реализовать

So1ovyov commented on Tracker/Domain/DataManager.swift:
```diff
@@ -0,0 +1,136 @@
+        deletedIndexes = nil
+    }
+    
+    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
```
Давай длинные строки распределять компактнее а не растягивать в одну строку, так будет легче читать код:
```
    func controller(
        _ controller: NSFetchedResultsController<NSFetchRequestResult>,
        didChange anObject: Any,
        at indexPath: IndexPath?,
        for type: NSFetchedResultsChangeType,
        newIndexPath: IndexPath?
    ) {
```

So1ovyov commented on Tracker/Domain/DataManager.swift:
```diff
@@ -0,0 +1,136 @@
+    func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChange anObject: Any, at indexPath: IndexPath?, for type: NSFetchedResultsChangeType, newIndexPath: IndexPath?) {
+        
+        switch type {
+        case .delete:
```
Кажется тут повторяется логика давай сократим код:
```
        switch type {
        case .insert, .delete:
            delegate?.didUpdate()
        default:
            break
        }
```

So1ovyov commented on Tracker/Helpers/UIColor+Extensions.swift:
```diff
@@ -0,0 +1,49 @@
+    }
+}
+
+//extension UIColor {
```
Кажется это тестовый код, который уже не пригодился (реализация рабочего кода выше как я понял)
Давай удалим )

So1ovyov commented on Tracker/View/HabitView/NewHabitViewController.swift:
```diff
@@ -353,11 +353,11 @@ final class NewHabitViewController: UIViewController {
-    private func checkIfAllFieldsFilledOut() -> Bool {
+    private func allFieldsFilledOut() -> Bool {
         if typeTracker == .irregularIvent {
             schedule = Schedule.allCases.map({ $0 })
```
Записи такого типа не обязательно круглыми скобками обводить, вот такая запись валидная и проще
`schedule = Schedule.allCases.map { $0 }`

So1ovyov commented on Tracker/View/HomeView/HomeViewConrtoller.swift:
```diff
@@ -20,11 +20,18 @@ final class HomeViewController: UIViewController {
+    private var categories: [TrackerCategoryProtocol] = []
+    private var visibleCategories: [TrackerCategoryProtocol] = []
+    private lazy var dataManager: DataManagerProtocol? = {
+        let dataStore = DataStore()
```
Может разгрузим это свойство и всю логику ниже вынесем в какой-нибудь приватный метод ?

Kolesnikov-Eugene replied:
Создал отдельный метод настройки для свойства.

So1ovyov commented on Tracker/View/HomeView/HomeViewConrtoller.swift:
```diff
@@ -95,6 +102,9 @@ final class HomeViewController: UIViewController {
         addSubviews()
         applyConstraints()
+        
+        categories = try! dataManager?.fetchAllCategories() ?? []
```
Force try плохая практика, давай используем try?  и guard-ом обезопасим себя.

Kolesnikov-Eugene replied:
Заменил Force unwrap в проекте, вроде бы больше не осталось)

So1ovyov commented on Tracker/View/HomeView/HomeViewConrtoller.swift:
```diff
@@ -341,21 +347,19 @@ extension HomeViewController: HomeViewCellDelegate {
+extension HomeViewController: DataManagerDelegate {
+    func didUpdate() {
+        categories.removeAll()
+        categories = try! dataManager?.fetchAllCategories() ?? []
```
Опять force try, давай избавимся.

So1ovyov commented on Tracker/Core Data/NullStore.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+import CoreData
+
+final class NullStore {}
```
Название как-то не говорит для чего этот класс. Может ManagerStore какой-нибудь ?

So1ovyov replied:
Я не совсем понимаю для чего этот класс ? Ты его нигде не используешь в проекте. Может это устаревший код который не нужен ?

Kolesnikov-Eugene replied:
Забыл его удалить, он уже не используется в проекте.

So1ovyov commented on Tracker/Core Data/DataStore.swift:
```diff
@@ -0,0 +1,195 @@
+//MARK: - TrackerCategoryStore protocol
+extension DataStore: TrackerCategoryStore {
+    func deleteCategory(_ category: NSManagedObject) throws {
+        //TODO later?
```
Может добавим подробное описание что будет в следующей итерации ?

So1ovyov commented on Tracker/Core Data/DataStore.swift:
```diff
@@ -0,0 +1,195 @@
+        try context.save()
+    }
+    
+    func deleteTracker(_ tracker: NSManagedObject) throws {
```
Давай на все пустые методы добавим тудушки и опишем что будет тут в будущем

So1ovyov commented on Tracker/Core Data/DataStore.swift:
```diff
@@ -0,0 +1,195 @@
+        let trackersData = try context.fetch(request)
+        var trackers = [TrackerProtocol]()
+        trackersData.forEach { tracker in
+            if let id = tracker.id,
```
Тут очень много вложенности, в if есть guard, давай декомпозируем как-то

Kolesnikov-Eugene replied:
Постарался декомпозировать. Теперь вроде получше выглядит)

Kolesnikov-Eugene committed:
removed class NullStrore. It is not used in project anymore.
Changed all logic where force unwrap used (in HomeViewController and DataStore).
Commented unused code removed from the project.
Added TODO descriptions for future methods (for Sprint 16).
Refactoring
